File: app/src/test/java/com/example/fruitsdiary/DiaryPresenterTest.java
package com.example.fruitsdiary;

import com.example.fruitsdiary.data.entry.EntryRepository;
import com.example.fruitsdiary.exception.FruitDiaryException;
import com.example.fruitsdiary.model.Entry;
import com.example.fruitsdiary.usecase.diary.DiaryContract;
import com.example.fruitsdiary.usecase.diary.DiaryPresenter;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.Mockito;
import org.mockito.runners.MockitoJUnitRunner;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.Callable;

import io.reactivex.Observable;
import io.reactivex.ObservableEmitter;
import io.reactivex.ObservableOnSubscribe;

import static org.junit.Assert.assertEquals;

/**
 * Example local unit test, which will execute on the development machine (host).
 *
 * @see <a href="http://d.android.com/tools/testing">Testing documentation</a>
 */

@RunWith(MockitoJUnitRunner.class)
public class DiaryPresenterTest {

    @Mock
    EntryRepository mEntryRepository;

    private DiaryPresenter mPresenter;

    @Mock
    private DiaryContract.View mView;

    @Before
    public void setUp() throws Exception {
        mPresenter = new DiaryPresenter(mEntryRepository);
        mPresenter.setView(mView);
    }

    @Test
    public void populateEntriesSuccess() {
        Mockito.when(mEntryRepository.getAllEntries()).thenReturn(Observable.just(getMockEntryList()));
        mPresenter.populateEntries();
        Mockito.verify(mView).showEntries(Mockito.anyList());
    }

    @Test
    public void populateEntriesFailure() {
        Mockito.when(mEntryRepository.getAllEntries()).thenReturn(Observable.<List<Entry>>error(new Throwable()));
        mPresenter.populateEntries();
        Mockito.verify(mView).handleNetworkError(Mockito.any(FruitDiaryException.class));
    }

    @After
    public void tearDown() throws Exception {
        mPresenter.setView(null);
    }

    private List<Entry> getMockEntryList(){
        List<Entry> entryList = new ArrayList<>();
        Entry entry1 = new Entry();
        Entry entry2 = new Entry();
        Entry entry3 = new Entry();
        entry1.setDate("2019-12-02");
        entry2.setDate("2019-12-03");
        entry2.setDate("2019-12-04");
        entryList.add(entry1);
        entryList.add(entry2);
        entryList.add(entry3);
        return entryList;
    }
}

File: app/src/test/java/com/example/fruitsdiary/DiaryPresenterTest.kt
package com.example.fruitsdiary

import com.example.fruitsdiary.data.entry.EntryRepository
import com.example.fruitsdiary.exception.FruitDiaryException
import com.example.fruitsdiary.model.Entry
import com.example.fruitsdiary.usecase.diary.DiaryContract
import com.example.fruitsdiary.usecase.diary.DiaryPresenter
import io.reactivex.Observable
import org.junit.After
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.Mock
import org.mockito.Mockito
import org.mockito.runners.MockitoJUnitRunner
import java.util.*

/**
 * Example local unit test, which will execute on the development machine (host).
 *
 * @see [Testing documentation](http://d.android.com/tools/testing)
 */

@RunWith(MockitoJUnitRunner::class)
class DiaryPresenterTest {

    @Mock
    internal var mEntryRepository: EntryRepository? = null

    private var mPresenter: DiaryPresenter = DiaryPresenter(mEntryRepository)

    @Mock
    private val mView: DiaryContract.View? = null

    private val mockEntryList: List<Entry>
        get() {
            val entryList = ArrayList<Entry>()
            val entry1 = Entry()
            val entry2 = Entry()
            val entry3 = Entry()
            entry1.date = "2019-12-02"
            entry2.date = "2019-12-03"
            entry2.date = "2019-12-04"
            entryList.add(entry1)
            entryList.add(entry2)
            entryList.add(entry3)
            return entryList
        }

    @Before
    @Throws(Exception::class)
    fun setUp() {
        mPresenter = DiaryPresenter(mEntryRepository)
        mPresenter.setView(mView)
    }

    @Test
    fun populateEntriesSuccess() {
        Mockito.`when`(mEntryRepository!!.allEntries).thenReturn(Observable.just(mockEntryList))
        mPresenter.populateEntries()
        Mockito.verify<DiaryContract.View>(mView).showEntries(Mockito.anyListOf(Entry::class.java))
    }

    @Test
    fun populateEntriesFailure() {
        Mockito.`when`(mEntryRepository!!.allEntries).thenReturn(Observable.error(Throwable()))
        mPresenter.populateEntries()
        Mockito.verify<DiaryContract.View>(mView).handleNetworkError(Mockito.any<FruitDiaryException>(FruitDiaryException::class.java))
    }

    @After
    @Throws(Exception::class)
    fun tearDown() {
        mPresenter.setView(null)
    }
}